hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/IncompatibleMutationAdapter.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.bigtable.hbase.replication.adapters;
+
+import com.google.cloud.bigtable.hbase.adapters.DeleteAdapter;
+import com.google.cloud.bigtable.hbase.replication.metrics.MetricsExporter;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.Cell;
+import org.apache.hadoop.hbase.CellUtil;
+import org.apache.hadoop.hbase.KeyValue;
+import org.apache.hadoop.hbase.client.Connection;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Base class for adapters that translate CBT incompatible mutations into compatible mutations. See
+ * <a href="https://cloud.google.com/bigtable/docs/hbase-differences#mutations_and_deletions">cbt
+ * docs</a> for detailed list of incompatible mutations.
+ *
+ * <p>Subclasses must expose the constructor ChildClass(Configuration, MetricSource, Table).
+ */
+public abstract class IncompatibleMutationAdapter {
+
+  private static final Logger LOG = LoggerFactory.getLogger(IncompatibleMutationAdapter.class);
+  // Connection to CBT. This can be used by child classes to fetch current state of row from CBT.
+  // For example: DeleteFamilyVersion can be implemented by fetching all the cells for the version
+  // (using this connection) and then deleting them.
+  private final Connection connection;
+  private final Configuration conf;
+  private final MetricsExporter metricsExporter;
+
+  public static final String INCOMPATIBLE_MUTATION_METRIC_KEY = "bigtableIncompatibleMutations";
```

Comment:
Consider moving all of the metrics behind a a class so that they are centralized and typed